overlord/snapstate/autorefresh_gating.go

```diff
@@ -32,6 +35,249 @@ import (
+	}
+	st.Set("refresh-gating", gating)
+	if len(herr.SnapsInError) > 0 {
+		return herr
+	}
+	return nil
+}
+
+// ProceedWithRefresh unblocks all snaps held by gatingSnap for refresh. This
+// should be called for --proceed on the gatingSnap.
+func ProceedWithRefresh(st *state.State, gatingSnap string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	if len(gating) == 0 {
+		return nil
+	}
+
+	var changed bool
+	for affectingSnap, gatingSnaps := range gating {
+		if _, ok := gatingSnaps[gatingSnap]; ok {
+			delete(gatingSnaps, gatingSnap)
+			changed = true
+		}
+		if len(gating[affectingSnap]) == 0 {
+			delete(gating, affectingSnap)
+		}
+	}
+
+	if changed {
+		st.Set("refresh-gating", gating)
+	}
+	return nil
+}
+
+// resetGating resets gating information by:
+// - removing affecting snaps whose held time expired.
+// - removing affecting snaps that are not in candidates (meaning there is no update for them anymore).
+func resetGating(st *state.State, candidates map[string]*refreshCandidate) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+
+	// optimize some edge cases
+	if len(gating) == 0 {
+		return nil
+	}
+	if len(candidates) == 0 {
+		gating = map[string]map[string]*holdInfo{}
+		st.Set("refresh-gating", gating)
+		return nil
+	}
+
+	now := timeNow()
+
+	var changed bool
+	for affectingSnap, gatingSnaps := range gating {
+		if candidates[affectingSnap] == nil {
+			// the snap doesn't have an update anymore, forget it
+			delete(gating, affectingSnap)
+			changed = true
+			continue
+		}
+
+		for gatingSnap, tm := range gatingSnaps {
+			if tm.HoldUntil.Before(now) {
+				delete(gatingSnaps, gatingSnap)
+				changed = true
+			}
+		}
+		// after deleting gating snaps we may end up with empty map under
+		// gating[affectingSnap], so remove it.
+		if len(gatingSnaps) == 0 {
+			delete(gating, affectingSnap)
+		}
+	}
+	if changed {
+		st.Set("refresh-gating", gating)
+	}
+	return nil
+}
+
+// resetGatingForRefreshed resets gating information by removing refreshedSnaps
+// (they are not held anymore). This should be called for all successfully
+// refreshed snaps.
+func resetGatingForRefreshed(st *state.State, refreshedSnaps ...string) error {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return err
+	}
+	if len(gating) == 0 {
+		return nil
+	}
+
+	var changed bool
+	for _, snapName := range refreshedSnaps {
+		if _, ok := gating[snapName]; ok {
+			delete(gating, snapName)
+			changed = true
+		}
+	}
+
+	if changed {
+		st.Set("refresh-gating", gating)
+	}
+	return nil
+}
+
+// heldSnaps returns all snaps that are gated and shouldn't be refreshed.
+func heldSnaps(st *state.State) (map[string]bool, error) {
+	gating, err := refreshGating(st)
+	if err != nil {
+		return nil, err
+	}
+	if len(gating) == 0 {
+		return nil, nil
+	}
+
+	held := make(map[string]bool)
+	for affecting, gatingSnaps := range gating {
+		if len(gatingSnaps) > 0 {
+			held[affecting] = true
```

Comment:
shouldn't this consider the times in holdInfo ?

Reply:
It can, my idea was to expire the entries early in auto-refresh, then heldSnaps() will be consulted later during same auto-refresh (the information from heldSnaps would of course be timely only shortly after we prune snaps-hold map).

Reply:
this was done now I see?